# Embedding techniques

In [ ]:
import os
from dotenv import load_dotenv
load_dotenv() # Load all the environment variables

True

In [ ]:
# Copie la clé API depuis ton fichier .env vers les variables d’environnement du système Python
os.environ["OPENAI_API_KEY"]=os.getenv("OPENAI_API_KEY")

In [5]:
from langchain_openai import OpenAIEmbeddings

embeddings=OpenAIEmbeddings(model="text-embedding-3-large")
embeddings

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x000001B712571A50>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x000001B712571900>, model='text-embedding-3-large', dimensions=None, deployment='text-embedding-ada-002', openai_api_version=None, openai_api_base=None, openai_api_type=None, openai_proxy=None, embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

In [6]:
text = "This is a tutorial on OpenAI Embedding"

query_result=embeddings.embed_query(text)
query_result

[0.0024039242416620255,
 0.03108314797282219,
 -0.012268977239727974,
 -0.03874962776899338,
 0.02165650576353073,
 -0.004540564958006144,
 0.016441212967038155,
 0.05408259108662605,
 -0.013885718770325184,
 -0.0005480132531374693,
 -0.02444668672978878,
 0.0062974668107926846,
 -0.0018856545211747289,
 -0.00925714522600174,
 0.010815214365720749,
 0.014211674220860004,
 0.015658918768167496,
 0.04221779853105545,
 -0.03329964727163315,
 -0.01908797398209572,
 0.02366439253091812,
 0.014029138721525669,
 -0.04782424122095108,
 -0.04005345329642296,
 0.022673487663269043,
 -0.021930309012532234,
 0.015450306236743927,
 0.04818930849432945,
 -0.008448774926364422,
 0.05144886672496796,
 0.014302941970527172,
 -0.014837509952485561,
 0.001381237874738872,
 -0.010130707174539566,
 -0.02031356655061245,
 0.026232926174998283,
 0.0435737743973732,
 0.048058927059173584,
 -0.05340460315346718,
 0.0203526820987463,
 0.034212324768304825,
 0.009491833858191967,
 0.0030199808534234762,
 -0.0134

In [7]:
len(query_result)

3072

Ce vecteur d'une longueur de 3072 représente le sens global de la phrase

In [15]:
embeddings_1024=OpenAIEmbeddings(model="text-embedding-3-large",dimensions=1024)

text = "This is a tutorial on OpenAI Embedding"

query_result=embeddings_1024.embed_query(text)
query_result

[0.003313322551548481,
 0.04297388345003128,
 -0.017018020153045654,
 -0.053699299693107605,
 0.029973382130265236,
 -0.006288090720772743,
 0.022732824087142944,
 0.07489734143018723,
 -0.019121572375297546,
 -0.0007369208615273237,
 -0.033873531967401505,
 0.008725685067474842,
 -0.0026542693376541138,
 -0.012828968465328217,
 0.014896409586071968,
 0.019699372351169586,
 0.02166750468313694,
 0.05846615135669708,
 -0.04611567407846451,
 -0.026416299864649773,
 0.03280821442604065,
 0.019410472363233566,
 -0.06619422882795334,
 -0.055396586656570435,
 0.03139982745051384,
 -0.030406732112169266,
 0.02134249173104763,
 0.06680814176797867,
 -0.01168239675462246,
 0.07121386379003525,
 0.0198257677257061,
 -0.02054801769554615,
 0.0019703886937350035,
 -0.014065821655094624,
 -0.028167756274342537,
 0.03636529669165611,
 0.06034399941563606,
 0.06644701212644577,
 -0.0739223062992096,
 0.028239982202649117,
 0.04741572216153145,
 0.013099811971187592,
 0.004175508860498667,
 -0.0185979

In [16]:
len(query_result)

1024

In [13]:
# Text loader

from langchain_community.document_loaders import TextLoader

loader=TextLoader('speech.txt')

docs=loader.load()
docs

[Document(metadata={'source': 'speech.txt'}, page_content='The world must be made safe for democracy. Its peace must be planted upon the tested foundations of political liberty. We have no selfish ends to serve. We desire no conquest, no dominion. We seek no indemnities for ourselves, no material compensation for the sacrifices we shall freely make. We are but one of the champions of the rights of mankind. We shall be satisfied when those rights have been made as secure as the faith and the freedom of nations can make them.\n\nJust because we fight without rancor and without selfish object, seeking nothing for ourselves but what we shall wish to share with all free peoples, we shall, I feel confident, conduct our operations as belligerents without passion and ourselves observe with proud punctilio the principles of right and of fair play we profess to be fighting for.\n\n…\n\nIt will be all the easier for us to conduct ourselves as belligerents in a high spirit of right and fairness 

In [14]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(chunk_size=500,chunk_overlap=50) # L'overlap permet d'avoir un nombre de derniers caractères au début du groupe d'après
final_documents=text_splitter.split_documents(docs)
final_documents

[Document(metadata={'source': 'speech.txt'}, page_content='The world must be made safe for democracy. Its peace must be planted upon the tested foundations of political liberty. We have no selfish ends to serve. We desire no conquest, no dominion. We seek no indemnities for ourselves, no material compensation for the sacrifices we shall freely make. We are but one of the champions of the rights of mankind. We shall be satisfied when those rights have been made as secure as the faith and the freedom of nations can make them.'),
 Document(metadata={'source': 'speech.txt'}, page_content='Just because we fight without rancor and without selfish object, seeking nothing for ourselves but what we shall wish to share with all free peoples, we shall, I feel confident, conduct our operations as belligerents without passion and ourselves observe with proud punctilio the principles of right and of fair play we profess to be fighting for.\n\n…'),
 Document(metadata={'source': 'speech.txt'}, page_

In [17]:
# Vector Embedding  and vector storeDB

from langchain_community.vectorstores import Chroma

# embeddings_1024=OpenAIEmbeddings(model="text-embedding-3-large",dimensions=1024)

db = Chroma.from_documents(final_documents, embeddings_1024)

In [18]:
db

In [ ]:
# Retrieve the results from query vecrorstoredb
query = "It will be all the easier for us to conduct ourselves as belligerents"

retrieved_result = db.similarity_search(query)
print(retrieved_result)

[Document(metadata={'source': 'speech.txt'}, page_content='It will be all the easier for us to conduct ourselves as belligerents in a high spirit of right and fairness because we act without animus, not in enmity toward a people or with the desire to bring any injury or disadvantage upon them, but only in armed opposition to an irresponsible government which has thrown aside all considerations of humanity and of right and is running amuck. We are, let me say again, the sincere friends of the German people, and shall desire nothing so much as the early'), Document(metadata={'source': 'speech.txt'}, page_content='Just because we fight without rancor and without selfish object, seeking nothing for ourselves but what we shall wish to share with all free peoples, we shall, I feel confident, conduct our operations as belligerents without passion and ourselves observe with proud punctilio the principles of right and of fair play we profess to be fighting for.\n\n…'), Document(metadata={'sou